### ISSUE TITLE
Default github artifact does not set 'name' field

### ISSUE THREAD

**Issue by ezimanyi** [spinnaker/spinnaker#2572]
When creating a default artifact of type 'github', we provide fields for the user to supply the 'reference' and 'version' of the artifact, but do not provide a field for the file name.  This means that downstream checks that try to bind to this artifact will fail if they are matching by name.

We should parse the 'reference' field to determine the file name and populate it (similar to how we determine the name and version for a docker artifact).

**Comment by ezimanyi** [comment]
@skim1420 : This explains the issue you were observing earlier with your github artifacts not matching.

**Comment by ezimanyi** [comment]
I have a tentative fix coded that simply takes the input `Content URL` (stored in the `reference` field), parses out the file path, and sets that as the `File path` (stored in the `name` field) on the default artifact.

Confirmed that now that the default artifact has its `name` field properly set, it can satisfy downstream expected artifact constraints (which it was not able to do before).

**Comment by wjoel** [comment]
@ezimanyi we extract the name from Docker image references, perhaps you could use this functionality as well? See `getNameFromReference` and `namePattern` in https://github.com/spinnaker/clouddriver/blob/b8c0a49323fa3d984e49dcca597475440b5e84a0/clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/artifact/ArtifactReplacer.java#L151 and https://github.com/spinnaker/clouddriver/blob/master/clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/artifact/ArtifactReplacerFactory.java#L46

**Comment by skim1420** [comment]
There's a question of whether `Name` for artifacts should be agnostic to the artifact type, rather than being a strong hint of the artifact type. The motivation is reducing complexity or the number of things to think about with artifacts. Since `Type` and `Reference` both clearly specify and strongly hint at the type, there's no explicit need for `Name` to collude as well. Name is just an arbitrary string for identifying the artifact and operating on it in various contexts.

Any thoughts on this? Can we articulate use cases whether this would degrade the design, or examples where we care what `Name` means?

**Comment by wjoel** [comment]
For Docker images, the name is used in order to do the replacement in the Kubernetes manifest. The name of the artifact must match the `image:` field. It is also  extracted from existing resources, and for instance the name `nginx` is a sensible choice for a deployment with a container using `image: nginx:1.13.10-alpine`. Finally, it is also extracted from Docker hub events, where we need to pick something, and again the name of the image (without the tag) seems like a reasonable choice. An arbitrary name would not be beneficial.

For GitHub (this issue), a name needs to be extracted from a reference, and the path to the file seems like a reasonable choice. It also matches the names for artifacts received from GitHub webhooks.

For GCS, S3 and other storage that can send events (pubsub, SQS, etc.), either the key of the blob (`path/to/my/file.txt`) or the URL of the blob (`s3://my-bucket/path/to/my/file.txt`) might be good names.

While it's easy to decouple the name from the reference for static artifacts, it doesn't seem like a good idea in general. For example, I want `manifests/my-app/deployment.yaml` to be a stable name that I can use in the pipelines, no matter what the git commit hash is in the webhook that triggers my pipeline, and I want to be able to match on that name using an expected artifact. What makes a good name does and should depend on the type of the artifact, in my opinion.

**Comment by ezimanyi** [comment]
@wjoel : Thanks for pointing me at the code where we do this replacement for Docker images.  I'd actually been considering a front-end solution whereby we extract the artifact name from the reference at the time the artifact is saved in the UI, similar to what we do for docker artifacts [here](https://github.com/spinnaker/deck/blob/987d268481d7dace92cf3dc24a616c09fee1ce5e/app/scripts/modules/core/src/pipeline/config/triggers/artifacts/docker/defaultDocker.artifact.ts#L23).

It looks like the implementation of this replacement in the back-end is specifically to handle cases where we find an artifact in a manifest.  It looks like that code doesn't handle Github artifacts at all---presumably because we don't support including Github artifacts in a mainfest.  Am I reading that correctly, or is there some change that should still be made to that code?

**Comment by ezimanyi** [comment]
My change above spinnaker/deck#5083 doesn't address the broader question of how to handle the various artifact fields, it just fixes the fact that default Github artifacts are effectively completely broken by setting the name field when saving a default Github artifact.

I do think it's worth considering more broadly how we use the various artifact fields---and even if we don't make any changes to the data model, consider whether there are changes to what we surface in the UI that could be helpful.  Just as one example, even showing the deduced 'name' of an artifact as a non-editable field might make it a bit clearer how this functionality works.

**Comment by dchw** [comment]
I ran into this too, at one point. I had to specify a custom artifact to match a GitHub default artifact, and it was extremely non-obvious. I think your fix seem pragmatic and useful until this larger discussion is completed.

FWIW here is the custom match I had to do for a default GitHub artifact:
```
      "matchArtifact": {
        "kind": "custom",
        "reference": ".*base.yaml",
        "type": "github/file"
      },
```

**Comment by lwander** [comment]
> presumably because we don't support including Github artifacts in a manifest. Am I reading that correctly, or is there some change that should still be made to that code?

@ezimanyi that's right, it doesn't really make sense to bind a github file inside a Kubernetes manifest, at least not with the current semantics.

---

@dchw which version of Spinnaker are you running? If it's 1.6.1, @ezimanyi's fix isn't patched into that yet - but this should be working on the nightlies:

>  it just fixes the fact that default Github artifacts are effectively completely broken by setting the name field when saving a default Github artifact.

**Comment by lwander** [comment]
No worries! 

@ezimanyi can you patch your fix into `release-1.6.x`?

**Comment by ezimanyi** [comment]
Merged into `release-1.6.x` in https://github.com/spinnaker/deck/pull/5162

**Comment by lwander** [comment]
Closing as this will be available in 1.6.2
